### PR TITLE
Use new-style Faabric snapshots

### PR DESCRIPTION
**Pull request by Shillaker**
This updates Faasm to use the new RPC-based approach to snapshot and restore in Faabric.

Previously there was a lot of complexity in the existing wasm module around snapshotting and using CoW memory. I've tried to encapsulate all generic snapshot-related functionality in Faabric itself, and only have the `WasmModule` class know about snapshotting in Faasm. Having such a low-level class be aware of managing snapshot-and-restore seems counterintuitive, but actually it makes the code a lot simpler.

Release build cancelled as it was either taking too long or had hung.

### CODE REVIEW

File: src/faaslet/Faaslet.cpp [hunk 1]
```diff
@@ -118,14 +118,12 @@ void Faaslet::preFinishCall(faabric::Message& call,
...
-        module_cache::WasmModuleCache& registry =
-          module_cache::getWasmModuleCache();
-        wasm::WAVMWasmModule& cachedModule = registry.getCachedModule(call);
 
-        auto* wavmModulePtr = dynamic_cast<wasm::WAVMWasmModule*>(module.get());
-        *wavmModulePtr = cachedModule;
+        // Restore from zygote
+        wasm::WAVMWasmModule& cachedModule =
+          module_cache::getWasmModuleCache().getCachedModule(call);
+        module = std::make_unique<wasm::WAVMWasmModule>(cachedModule);
```
Comment (Shillaker) [inline src/faaslet/Faaslet.cpp:126]: This isn't doing anything different, just tidying up syntax

File: src/faaslet/Faaslet.cpp [hunk 2]
```diff
@@ -151,24 +149,10 @@ void Faaslet::postBind(const faabric::Message& msg, bool force)
...
-        wasm::WAVMWasmModule& snapshot = registry.getCachedModule(msg);
-
-        // Use snapshot to restore WAVM module
-        module = std::make_unique<wasm::WAVMWasmModule>(snapshot);
-
-        PROF_END(snapshotRestore)
+        // Get cached module
+        wasm::WAVMWasmModule& cachedModule =
+          module_cache::getWasmModuleCache().getCachedModule(msg);
+        module = std::make_unique<wasm::WAVMWasmModule>(cachedModule);
```
Comment (Shillaker) [inline src/faaslet/Faaslet.cpp:155]: Again, just tidying up syntax.

File: src/module_cache/WasmModuleCache.cpp [hunk 3]
```diff
@@ -24,75 +24,26 @@ int WasmModuleCache::getCachedModuleCount(const std::string& key)
...
+    // Check for the base cached module (must be present for restoring a
+    // snapshot too)
     if (getCachedModuleCount(baseKey) == 0) {
         faabric::util::FullLock lock(mx);
         if (cachedModuleMap.count(baseKey) == 0) {
             // Instantiate the base module
             logger->debug("Creating new base zygote: {}", baseKey);
             wasm::WAVMWasmModule& module = cachedModuleMap[baseKey];
             module.bindToFunction(msg);
-
```
Comment (Shillaker) [inline src/module_cache/WasmModuleCache.cpp:56]: The `WasmModuleCache` was aware of snapshot and restore previously, which seemed odd as its purpose is really just to cache the base cached version of each wasm module. Now it only ever holds a single cached module for each function.

File: src/wasm/WasmModule.cpp [hunk 4]
```diff
@@ -58,19 +60,28 @@ wasm::WasmEnvironment& WasmModule::getWasmEnvironment()
...
-    const std::shared_ptr<faabric::state::StateKeyValue>& stateKv =
-      state.getKV(getBoundUser(), stateKey, stateSize);
+    faabric::util::SnapshotData data = doSnapshot();
+    faabric::snapshot::SnapshotRegistry& reg =
+      faabric::snapshot::getSnapshotRegistry();
+    reg.setSnapshot(snapKey, data);
 
-    stateKv->set(snapData.data());
-    stateKv->pushFull();
+    return snapKey;
```
Comment (Shillaker) [inline src/wasm/WasmModule.cpp:75]: Rather than using state this now uses the Faabric snapshot registry.

File: src/wavm/mpi.cpp [hunk 5]
```diff
@@ -1355,7 +1355,8 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(env,
     int* coordsArray =
       Runtime::memoryArrayPtr<int>(ctx.memory, (Uptr)coords, (Uptr)maxdims);
 
-    ctx.world.getCartesianRank(ctx.rank, dimsArray, periodsArray, coordsArray);
+    ctx.world.getCartesianRank(
+      ctx.rank, maxdims, dimsArray, periodsArray, coordsArray);
```
Comment (Shillaker) [inline src/wavm/mpi.cpp:1359]: Changes ported to allow compiling against latest Faabric. Should be able to revert once Faabric PR merged in.

File: tests/test/faaslet/test_mpi.cpp [hunk 6]
```diff
@@ -41,10 +41,10 @@ TEST_CASE("Test MPI cartesian create", "[mpi]")
...
 }
 
-TEST_CASE("Test MPI cartesian coordinates", "[mpi]")
-{
-    checkMpiFunc("mpi_cartesian");
-}
+// TEST_CASE("Test MPI cartesian coordinates", "[mpi]")
+//{
+//   checkMpiFunc("mpi_cartesian");
+//}
```
Comment (Shillaker) [inline tests/test/faaslet/test_mpi.cpp:47]: Temporarily commenting out until MPI-related PR merged

File: src/module_cache/WasmModuleCache.cpp [hunk 7]
```diff
@@ -24,75 +24,26 @@ int WasmModuleCache::getCachedModuleCount(const std::string& key)
     return count;
 }
 
-std::string WasmModuleCache::getBaseCachedModuleKey(const faabric::Message& msg)
-{
-    std::string key = msg.user() + "/" + msg.function();
-    return key;
-}
```
Comment (Shillaker) [inline src/module_cache/WasmModuleCache.cpp:31]: This is basically `faabric::util::funcToString`

File: src/module_cache/WasmModuleCache.cpp [hunk 8]
```diff
@@ -24,75 +24,26 @@ int WasmModuleCache::getCachedModuleCount(const std::string& key)
...
-
-/**
- * There are two kinds of cached module here, the "base" cached module, i.e. the
- * default module with its zygote function executed, (same for all instances),
- * or one of many "special" cached modules, those restored from snapshots
- * captured at arbitrary points (e.g. when spawning a thread).
- */
```
Comment (Shillaker) [inline src/module_cache/WasmModuleCache.cpp:38]: The `WasmModuleCache` was aware of snapshots, which added needless complexity and misdirection. Instead it now just holds a single cached module for each function, and knows nothing about any extra snapshotting that goes on _after_ that module has been used to clone others.

File: include/wasm/WasmModule.h [hunk 9]
```diff
@@ -100,23 +101,10 @@ class WasmModule
...
-
-    size_t snapshotToState(const std::string& stateKey);
+    std::string snapshot();
 
-    void restoreFromFile(const std::string& filePath);
-
-    void restoreFromMemory(const std::vector<uint8_t>& data);
-
-    void restoreFromState(const std::string& stateKey, size_t stateSize);
+    void restore(const std::string& snapshotKey);
```
Comment (Shillaker) [inline include/wasm/WasmModule.h:107]: Previously there were three forms of snapshot and restore, to files, memory and state. Now we just have the one (to memory)

File: src/wasm/WasmModule.cpp [hunk 10]
```diff
@@ -322,26 +300,6 @@ void WasmModule::writeWasmEnvToMemory(uint32_t envPointers, uint32_t envBuffer)
...
-
-void WasmModule::doSnapshot(std::ostream& outStream)
-{
-    throw std::runtime_error("doSnapshot not implemented");
-}
-
-void WasmModule::doRestore(std::istream& inStream)
-{
-    throw std::runtime_error("doRestore not implemented");
-}
```
Comment (Shillaker) [inline src/wasm/WasmModule.cpp:343]: Submodules of `WasmModule` (i.e. `WAMRWasmModule` and `WAVMWasmModule`) no longer need to provide custom snapshot/ restore behaviour, they just need to be able to provide a pointer to their memory array.

File: src/wasm/WasmModule.cpp [hunk 11]
```diff
@@ -58,77 +60,53 @@ wasm::WasmEnvironment& WasmModule::getWasmEnvironment()
...
 
-size_t WasmModule::snapshotToState(const std::string& stateKey)
+std::string WasmModule::snapshot()
 {
-    const std::vector<uint8_t> snapData = snapshotToMemory();
-    unsigned long stateSize = snapData.size();
-
-    faabric::state::State& state = faabric::state::getGlobalState();
-    const std::shared_ptr<faabric::state::StateKeyValue>& stateKv =
-      state.getKV(getBoundUser(), stateKey, stateSize);
```
Comment (Shillaker) [inline src/wasm/WasmModule.cpp:68]: Rather than using state this now uses the Faabric snapshot registry.

File: src/wavm/mpi.cpp (hunk 5 above)
Comment (csegarragonz) [reply src/wavm/mpi.cpp:1359]: Sure! I am working on it now.

**Pull request merged**